src/libmcrx.c

```diff
@@ -1055,3 +1076,418 @@ mcrx_is_system_error(enum mcrx_error_code err) {
+      struct mcrx_mnat_entry *new_entry = mcrx_mnat_ctx_find_entry_from_subscription(sub, clone_context);
+      if (sub->joined) {
+        if (new_entry == NULL || !mcrx_mnat_ctx_entry_local_equal(sub->mnat_entry, new_entry)) {
+          mcrx_subscription_leave(sub);
+          sub->mnat_entry = new_entry;
+          if (sub->mnat_entry
+              && !mcrx_mnat_ctx_entry_unresolved(sub->mnat_entry)) {
+            // rejoin with new entry
+            enum mcrx_error_code err = mcrx_subscription_native_join(sub);
+            if (err == MCRX_ERR_OK) {
+              ctx->live_subs++;
+            }
+          }
+        }
+      } else {
+        sub->mnat_entry = new_entry;
+        if (sub->mnat_entry && !mcrx_mnat_ctx_entry_unresolved(sub->mnat_entry)) {
+          enum mcrx_error_code err = mcrx_subscription_native_join(sub);
+          if (err == MCRX_ERR_OK) {
+            ctx->live_subs++;
+          }
+        }
+      }
+    }
+    // replace the run time mnat context
+    mcrx_mnat_ctx_unref(ctx->mnat_ctx);
+    ctx->mnat_ctx = clone_context;
+  }
+
+  info(ctx, "mnat context %p apply\n", (void* )mnatctx);
+
+  return MCRX_ERR_OK;
+}
+
+/**
+ * mcrx_mnat_ctx_find_entry:
+ *
+ * Retrieve the mcrx mnat entry based on global source and group addresses.
+ *
+ **/
+MCRX_EXPORT struct mcrx_mnat_entry* mcrx_mnat_ctx_find_entry(
+    struct mcrx_mnat_ctx *mnatctx, const char *global_source,
+    const char *global_group) {
+  struct mcrx_mnat_entry *entry = NULL;
+  if (global_source == NULL || global_group == NULL) {
+    return entry;
+  }
+  LIST_FOREACH(entry, &mnatctx->mnats_head, mnat_entries) {
+    int af;
+    const void *src, *grp;
+    char src_buf[INET6_ADDRSTRLEN];
+    char grp_buf[INET6_ADDRSTRLEN];
+    switch (entry->addr_type) {
+    case MCRX_ADDR_TYPE_V4:
+      af = AF_INET;
+      src = &entry->global_addrs.v4.source;
+      grp = &entry->global_addrs.v4.group;
+      break;
+    case MCRX_ADDR_TYPE_V6:
+      af = AF_INET6;
+      src = &entry->global_addrs.v6.source;
+      grp = &entry->global_addrs.v6.group;
+      break;
+    default:
+      continue;
+    }
+    const char *src_str = inet_ntop(af, src, src_buf, sizeof(src_buf));
+    if (src_str == NULL) {
+      continue;
+    }
+    const char *grp_str = inet_ntop(af, grp, grp_buf, sizeof(grp_buf));
+    if (grp_str == NULL) {
+      continue;
+    }
+    if (strcmp(src_str, global_source) == 0
+        && strcmp(grp_str, global_group) == 0) {
+      return entry;
+    }
+  }
+  return entry;
+}
+
+/**
+ * mcrx_mnat_ctx_add_entry:
+ *
+ * Note: unresolved local source and local group needed to be populated as 0.0.0.0
+ * Add the mcrx mnat entry.
+ *
+ **/
+MCRX_EXPORT enum mcrx_error_code mcrx_mnat_ctx_add_entry(
+    struct mcrx_mnat_ctx *mnatctx, const char *global_source,
+    const char *global_group, const char *local_source, const char *local_group) {
+  if (global_source == NULL || global_group == NULL || local_source == NULL
```

Comment:
I think we should consider refactoring this to separate the address-parsing into a different function call, and here accept a pre-parsed (S,G) object instead of strings.
There's several other problems with this function, for example it's an error if the local source and group address families don't match, or if the global source and group address families don't match, and the pton has to allow AF_INET6 instead of only accepting AF_INET.  I think it would be useful to separate that checking into a mcrx_sg_ntop function similar to the current subscription_config_ntop, and here work with a pre-parsed sg object.